### PR TITLE
fix: adjust typing of the parsers and compilers

### PR DESCRIPTION
**Pull request by johnhooks**
This commit adds missing types and attempts fix type errors.

There are still a few type errors, though how to fix them is not clear. Adds the `Translations` type for the `translations` property of the `GetTextTranslations` type.

### CODE REVIEW

File: src/mocompiler.js [hunk 1]
```diff
@@ -1,15 +1,31 @@
...
-import { Transform } from 'node:stream';
 
-/** @typedef {{msgid: number, msgstr: number, total: number}} Size data of {msgid, msgstr, total} */
+/**
+ * @typedef {import('node:stream').Transform} Transform
+ * @typedef {import('./types.js').GetTextTranslation} GetTextTranslation
+ * @typedef {import('./types.js').GetTextTranslations} GetTextTranslations
+ * @typedef {import('./types.js').Translations} Translations
+ * @typedef {import('./types.js').WriteFunc} WriteFunc
+ */
```
Comment (johnhooks) [inline src/mocompiler.js:11]: @erikyo I don't know if this was how you were importing them, but if this doesn't work for you I can switch back to inline importing at every instance.

File: src/mocompiler.js [hunk 2]
```diff
@@ -61,24 +76,24 @@ function prepareTranslations (translations) {
       }
 
       return result;
-    }, {});
+    }, /** @type {Record<string, GetTextTranslation>} */({}));
```
Comment (johnhooks) [inline src/mocompiler.js:79]: Declaring the type of the reducing value at the source.

File: src/pocompiler.js [hunk 3]
```diff
@@ -36,15 +43,15 @@ export function preparePoHeaders (headersRaw) {
...
 }
 
 /**
  * Creates a PO compiler object.
  *
  * @constructor
- * @param {import('./types.js').GetTextTranslations} [table] Translation table to be compiled
- * @param {Partial<import('./types.js').parserOptions>} [options] Options
+ * @param {GetTextTranslations} [table] Translation table to be compiled
+ * @param {ParserOptions} [options] Options
```
Comment (johnhooks) [inline src/pocompiler.js:58]: I capitalized it, also it doesn't really have to be marked as `Partial` it has no required properties.

File: src/pocompiler.js [hunk 4]
```diff
@@ -286,16 +293,13 @@ Compiler.prototype.compile = function () {
...
+  /** @type {GetTextTranslation} headerBlock */
   const headerBlock = (this._table.translations[''] && this._table.translations['']['']) || {};
 
-  /** @type {import('./types.js').GetTextTranslation[]|undefined} translations Prepared array */
   const translations = this._prepareSection(this._table.translations);
-  /** @type {string[]|undefined} response Prepared array */
   let response = translations?.map(t => this._drawBlock(t));
 
   if (typeof this._table.obsolete === 'object') {
-    /** @type {import('./types.js').GetTextTranslation[]|undefined} obsolete Prepared array */
```
Comment (johnhooks) [inline src/pocompiler.js]: These were not necessary, they are inferred.

File: src/poparser.js [hunk 5]
```diff
@@ -4,18 +4,29 @@ import { Transform } from 'readable-stream';
...
+ */
+
+/**
+ * @typedef {Object} Node A single Node object in the PO file
+ * @property {string} [key]
+ * @property {number} [type]
+ * @property {string} value
+ * @property {string} [quote]
+ * @property {boolean} [obsolete]
+ * @property {GetTextComment | undefined} [comments]
```
Comment (johnhooks) [inline src/poparser.js:31]: All the other interface types are declared this way, so I changes this one to match.

File: src/poparser.js [hunk 6]
```diff
@@ -169,14 +180,14 @@ Parser.prototype._lexer = function (chunk) {
             value: '',
             quote: chr
           };
-          this._lex.push(this._node);
+          this._lex.push(/** @type {Node} */ (this._node));
```
Comment (johnhooks) [inline src/poparser.js:185]: They start off as a `Partial<Node>`, but when they are pushed into the `_lex` they need to be forced into the correct type. This in JSDocs `/** @type {Node} */ (this._node)` is the same as `this._node as Node`.

File: src/poparser.js [hunk 7]
```diff
@@ -328,13 +338,13 @@ Parser.prototype._parseComments = function (tokens) {
 /**
  * Join gettext keys with values
  *
- * @param {Node[]& {key: { value: string } } } tokens - Parsed tokens containing key-value pairs
+ * @param {(Node & { value?: string })[]} tokens - Parsed tokens containing key-value pairs
```
Comment (johnhooks) [inline src/poparser.js:345]: I didn't see that this was wrapped in another object with a `key` property.

File: src/shared.js [hunk 8]
```diff
@@ -150,8 +150,9 @@ export function foldLine (str, maxLen = 76) {
 /**
  * Comparator function for comparing msgid
  *
- * @param {{msgid: string}} left with msgid prev
- * @param {{msgid: string}} right with msgid next
+ * @template {Buffer|string} T
+ * @param {{msgid: T}} left with msgid prev
+ * @param {{msgid: T}} right with msgid next
```
Comment (johnhooks) [inline src/shared.js:155]: This one actually is used on both strings and buffer, though only to compare a single type at a time.

File: src/types.js [hunk 9]
```diff
@@ -14,27 +14,39 @@
...
+ * @typedef {Record<string, Record<string, GetTextTranslation>>} Translations The translations index.
  */
 
 /**
  * Represents GetText translations.
  * @typedef {Object} GetTextTranslations
  * @property {string|undefined} charset Character set.
- * @property {{[key: string]: string}} headers Headers.
- * @property {GetTextTranslations['translations']} [obsolete] Obsolete messages.
- * @property {Object.<[msgctxt: string], Object.<[msgid: string], GetTextTranslation>>} translations Translations.
```
Comment (johnhooks) [inline src/types.js:28]: I didn't understand this type, it was coming up as `any`. Though I typed the `Translations` and used it.

File: src/mocompiler.js (hunk 1 above)
Comment (erikyo) [reply src/mocompiler.js:11]: I confirm that it works. It is not very nice to have a comment block at the beginning of each file, but that is how jsdocs are ¯\_(ツ)_/¯

File: src/poparser.js (hunk 6 above)
Comment (erikyo) [reply src/poparser.js:185]: ~isn't enough to declare `this._lex` as `node[]`?~ 

ah ok it was a partial before that

File: src/types.js (hunk 9 above)
Comment (erikyo) [reply src/types.js:28]: if can help, should type this

```
  "msgctxt": {
      "msgid": {
          "msgid": "",
          "msgstr": [
              "test"
          ]
      },
      "2": {
          "2": "",
          "msgstr": [
              "test"
          ]
      }
  }
```

File: src/mocompiler.js (hunk 1 above)
Comment (johnhooks) [reply src/mocompiler.js:11]: This has been my life. I just see it as a type import now.

File: src/types.js (hunk 9 above)
Comment (johnhooks) [reply src/types.js:28]: That what I typed it as, I got it from the docs. 

What I meant, I didn't understand the original type, did it work for you? I'm just curious, I had never seen that syntax before, the `Object.` in the `translation` property.

**Comment by erikyo** [comment]
Great job, @johnhooks! However, it seems there are still some (pregress) glitches resulting from mixed data types—strings occasionally becoming arrays and vice versa. In such instances, we might consider creating separate variables with distinct types to maintain clarity and consistency.

For instance, this issue arises when the translation object undergoes transformation into a text file, where all tokens are stringified and merged into the translation file indiscriminately. To address this, we could explore storing the resultant data elsewhere, thereby ensuring proper type checking is upheld. ([this ](https://github.com/smhg/gettext-parser/blob/6b3b940741e46564e199078876ed60fc4ea9747b/src/poparser.js#L330) is an example of mixed types)

As we tackle this, let's keep in mind the open issues. Some have already been reviewed, and closing them should be relatively straightforward.

File: src/poparser.js [hunk 10]
```diff
@@ -4,18 +4,31 @@ import { Transform } from 'readable-stream';
 import util from 'util';
 
 /**
- * Po parser options
+ * @typedef {import('stream').Stream.Writable} WritableState
```
Comment (erikyo) [inline src/poparser.js:7]: Why not to create a base Option and then extend that type to `PoParserOptions` and `MoParserOptions`?

**Comment by johnhooks** [comment]
@erikyo yes that was one of the instances I was having trouble grasping, when and where `node.value` is a `string` or a `Record<string, string>`.

File: src/poparser.js (hunk 10 above)
Comment (johnhooks) [reply src/poparser.js:7]: I'll look into that, I hadn't taken the time to compare them.

Comment (johnhooks) [reply src/poparser.js:7]: I looked into it, but the `Parser` in `moparser.js` doesn't take an options object. Can you clarify what you mean?

Comment (erikyo) [reply src/poparser.js:7]: OK I thought moParser had an options object like poParser has, and I suggested to standardise (but actually as you say MoParser does not have it and only has one string so the issue does not apply).

the one that is (maybe) slightly different is 
`{TransformOptions & {initialTreshold?: number;}} transformOptions`

that is the options set for the stream ([here](https://github.com/smhg/gettext-parser/blob/6b3b940741e46564e199078876ed60fc4ea9747b/src/poparser.js#L571)) but adding with `&` solves the issue for the moment

**Comment by erikyo** [comment]
to me it's all perfect! the only thing that i don't like too much, but which are probably necessary, are:

- ~Types output to a separate folder (i prefer to emit types in the same folder of the library). I am not aware of a difference between issuing types in ./types or directly into ./lib, and I see that modern bundlers by default use the second of these modes
it is probably a matter of preference rather than a real "need"~ ./types/ was only used only for development so it is actually related to the gettext-parser repo only

- [add encoding declaration](https://github.com/erikyo/gettext-parser-next/pull/2/commits/c072f28b0f296481ac9790e8c6d1f404d90f04e4) made me cringe a bit. Why not ask to the mantainer to emit types for that repo? I think `tsc lib/encoding.js --outDir types --allowJs --declaration --emitDeclarationOnly` would be enough to solve. PR?

**Comment by johnhooks** [comment]
@erikyo we DMed about this, but I thought I would include it here for clarification 

> That declare module “encoding” should only be for local type checking.

**Pull request merged**